pdns/lua-record.cc

```diff
@@ -60,30 +60,145 @@ class IsUpOracle
+      setDown(cd);
+    }
+  }
+  void checkTCP(const CheckDesc& cd, const bool status, const bool first = false) {
+    try {
+      int timeout = 1;
+      if (cd.opts.count("timeout")) {
+        timeout = std::atoi(cd.opts.at("timeout").c_str());
+      }
+      Socket s(cd.rem.sin4.sin_family, SOCK_STREAM);
+      ComboAddress src;
+      s.setNonBlocking();
+      if (cd.opts.count("source")) {
+        src = ComboAddress(cd.opts.at("source"));
+        s.bind(src);
+      }
+      s.connect(cd.rem, timeout);
+      if (!status) {
+        g_log<<Logger::Warning<<"Lua record monitoring declaring TCP/IP "<<cd.rem.toStringWithPort()<<" ";
+        if(cd.opts.count("source"))
+          g_log<<"(source "<<src.toString()<<") ";
+        g_log<<"UP!"<<endl;
+      }
+      setUp(cd);
+    }
+    catch (const NetworkError& ne) {
+      if(status || first) {
+        g_log<<Logger::Warning<<"Lua record monitoring declaring TCP/IP "<<cd.rem.toStringWithPort()<<" DOWN: "<<ne.what()<<endl;
+      }
+      setDown(cd);
+    }
+  }
+  void checkThread()
   {
-    std::thread thr;
-    bool status;
-  };
+    while (true)
+    {
+      std::chrono::system_clock::time_point checkStart = std::chrono::system_clock::now();
+      std::vector<std::future<void>> results;
+      std::vector<CheckDesc> toDelete;
+      {
+        ReadLock lock{&d_lock}; // make sure there's no insertion
+        for (auto& it: d_statuses) {
+          auto& desc = it.first;
+          auto& state = it.second;
+
+          if (desc.url.empty()) { // TCP
+            results.push_back(std::async(std::launch::async, &IsUpOracle::checkTCP, this, desc, state->status.load(), state->first.load()));
+          } else { // URL
+            results.push_back(std::async(std::launch::async, &IsUpOracle::checkURL, this, desc, state->status.load(), state->first.load()));
+          }
+          if (std::chrono::system_clock::from_time_t(state->lastAccess) < (checkStart - std::chrono::seconds(g_luaHealthChecksExpireDelay))) {
+            toDelete.push_back(desc);
+          }
+        }
+      }
+      // we can release the lock as nothing will be deleted
+      for (auto& future: results) {
+        future.wait();
+      }
+      if (!toDelete.empty()) {
+        WriteLock lock{&d_lock};
+        for (auto& it: toDelete) {
+          d_statuses.erase(it);
+        }
+      }
+      std::this_thread::sleep_until(checkStart + std::chrono::seconds(g_luaHealthChecksInterval));
+    }
+  }
 
-  typedef map<CheckDesc, Checker> statuses_t;
+  typedef map<CheckDesc, std::unique_ptr<CheckState>> statuses_t;
   statuses_t d_statuses;
 
-  std::mutex d_mutex;
+  std::unique_ptr<std::thread> d_checkerThread;
 
   void setStatus(const CheckDesc& cd, bool status)
   {
-    std::lock_guard<std::mutex> l(d_mutex);
-    d_statuses[cd].status=status;
+    auto& state = d_statuses[cd];
```

Comment:
Doing the access here without a lock is a bit scary. We know the entry already exists but, unless i'm missing something, a new insertion might occur at the exact same time from `IsUpOracle::isUp()`, right?